linmod/utils.py

```diff
@@ -53,3 +59,127 @@ def pl_softmax(pl_expr):
+        names = []
+        ess = []
+        psrf = []
+
+        if len(v["n_eff"].shape) == 0:
+            names = [k]
+            names_nodim = [k]
+            ess = [float(v["n_eff"])]
+            psrf = [float(v["r_hat"])]
+        else:
+            array_indices = [
+                idx for idx in product(*map(range, v["n_eff"].shape))
+            ]
+            names = [
+                k + "[" + ",".join([str(i) for i in idx]) + "]"
+                for idx in array_indices
+            ]
+            names_nodim = [k] * len(array_indices)
+            ess = [float(v["n_eff"][idx]) for idx in array_indices]
+            psrf = [float(v["r_hat"][idx]) for idx in array_indices]
+
+        param = pl.DataFrame(
+            {
+                "param": names,
+                "param_no_dim": names_nodim,
+                "n_eff": ess,
+                "r_hat": psrf,
+            }
+        )
+
+        convergence = pl.concat([convergence, param])
+
+    nans = convergence.filter(
+        (pl.col("n_eff").is_nan()) | (pl.col("r_hat").is_nan())
+    )
+    if nans.shape[0] > 0:
+        nans = nans.filter(~pl.col("param_no_dim").is_in(ignore_nan_in))
+        if nans.shape[0] > 0:
+            bad = nans["param"].unique().to_list()
+            raise RuntimeError(
+                "Found unexpected NaN convergence values for parameters: "
+                + str(bad)
+            )
+        if drop_ignorable_nan:
+            convergence = convergence.filter(
+                (pl.col("n_eff").is_not_nan()) & (pl.col("r_hat").is_not_nan())
+            )
+
+    return convergence.drop("param_no_dim")
+
+
+def plot_convergence(mcmc: MCMC, params: Iterable[str]):
+    posterior = mcmc.get_samples(group_by_chain=True)
+    plots = []
+    for par in params:
+        if par.count("[") > 0:
+            dimless_par = par.split("[")[0]
+            indices = [int(i) for i in par.split("[")[1][:-1].split(",")]
+            index_tuple = tuple(
+                [*[slice(None), slice(None)], *[i for i in indices]]
+            )
+            chains = posterior[dimless_par][index_tuple]
+        else:
+            chains = posterior[par]
+
+        niter = chains.shape[1]
+        df = pl.concat(
+            [
+                pl.DataFrame(
+                    {
+                        "param": [par] * niter,
+                        "chain": [str(chain)] * niter,
+                        "iteration": list(range(niter)),
+                        "value": chains[chain].tolist(),
+                    }
+                )
+                for chain in range(chains.shape[0])
+            ]
+        )
+
+        plt = (
+            ggplot(df)
+            + geom_line(
+                aes(
+                    x="iteration",
+                    y="value",
+                    color="chain",
+                ),
+            )
+            + theme_bw(base_size=20)
+        )
+
+        plots.append(plt)
+
+    return plots
```

Comment:
I think these are awesome, but I almost want them to belong in `linmod.models` or maybe even a new module?

In my head, `utils` was mainly for one-liners that support other package functionality, whereas this is full-fledged functionality in its own right

Keeping it with `models` might make sense, since it keeps all the numpyro-dependent code in one place

Reply:
One could also argue that the plotting should go in visualization, but we can put it in `linmod.models` for now.

Reply:
In the long run, there are plenty of reasons we may well end up refactoring modules or making some of what is now a single script module into a folder of scripts.

Reply:
In the long run, there are plenty of reasons we may well end up refactoring modules or making some of what is now a single script module into a folder of scripts.